File: solutions/clojure/largest-series-product/5/src/largest_series_product.clj
(ns largest-series-product)

(defn error-message? [span in-str in-list]
  (cond (neg? span) "span must not be negative"
    (> span (count in-str) ) "span must not exceed string length" 
    (> (count in-str) (count in-list)) "digits input must only contain digits"
    :else nil))

(def get-list #(clojure.string/split % #"(?=[0-9])"))
(defn get-digits [in] (map #(Integer/parseInt %) in))

(defn largest-product [span input-str] 
  (let [input-list (get-list input-str)
        error (error-message? span input-str input-list)]
      (when (some? error) (throw (IllegalArgumentException. error)))
         (->> (get-digits input-list)
             (partition span 1 [0])
             (map #(reduce * %))
             (apply max))))

File: solutions/python/largest-series-product/2/largest_series_product.py
from functools import reduce

def sub_product(substr):
    return reduce(lambda x,y: x*y, [int(c) for c in substr])

def only_digits(str):
    return len(list(filter(lambda c: c.isdigit(), str)))==len(str)

def largest_product(series, size):
    if size>len(series):
        raise ValueError("span must not exceed string length")
    elif size<0:
        raise ValueError("span must not be negative")
    elif not only_digits(series):
        raise ValueError("digits input must only contain digits")
    return max([sub_product(series[i:i+size]) for i in range(0,len(series)-size+1)])